### PR TITLE
feat(fluent source)!: Add log namespace and schema def support

### PR DESCRIPTION
**Pull request by neuronull**
Adds log namespace and schema def support as described in  #13720.

closes:  #15011
epic: #13723

### Reviewer notes:
- `amqp` and `aws_s3` were modified to include a "timestamp" to their schemas. This was discussed offline with Spencer when I noticed this missing from recent PRs.

### CODE REVIEW

File: src/sources/fluent/mod.rs [hunk 1]
```diff
@@ -92,8 +100,26 @@ impl SourceConfig for FluentConfig {
         )
     }
 
-    fn outputs(&self, _global_log_namespace: LogNamespace) -> Vec<Output> {
-        vec![Output::default(DataType::Log)]
+    fn outputs(&self, global_log_namespace: LogNamespace) -> Vec<Output> {
+        // `host_key` defaults to the `log_schema().host_key()` if it's not configured in the source.
+        let host_key = parse_value_path(log_schema().host_key())
```
[the remark this answers is not in the document]
Comment (neuronull) [reply src/sources/fluent/mod.rs:105]: Ah the comment in the code is incorrect, left from copy-paste.

The behavior here, I chose due what the code in `handle_events()` was doing- inserting a timestamp if the log doesn't already contain one.
I can fix that comment.

**Review by spencergilbert — DISMISSED** [review]
Adding this to the `fluent` code was less bad than I figured it was going to be 😄

File: src/sources/fluent/mod.rs [hunk 2]
```diff
@@ -445,16 +503,24 @@ impl From<FluentFrame> for SmallVec<[Event; 1]> {
...
+        log_namespace.insert_vector_metadata(
+            &mut log,
+            log_schema().timestamp_key(),
+            "ingest_timestamp",
+            timestamp,
+        );
+
         log.insert("tag", tag);
         for (key, value) in record.into_iter() {
             log.insert(event_path!(&key), value);
```
Comment (neuronull) [inline src/sources/fluent/mod.rs:526]: I guess these also need to be handled in the schema

Comment (spencergilbert) [reply src/sources/fluent/mod.rs:526]: Ah good catch!

Comment (neuronull) [reply src/sources/fluent/mod.rs:526]: though... the ones from `record`, are added dynamically and we don't know what the keys will be...

Comment (spencergilbert) [reply src/sources/fluent/mod.rs:526]: `aws_s3` has similarly unknown/dynamic entries: https://github.com/vectordotdev/vector/blob/master/src/sources/aws_s3/mod.rs#L161-L171

File: src/sources/fluent/mod.rs [hunk 3]
```diff
@@ -445,16 +503,24 @@ impl From<FluentFrame> for SmallVec<[Event; 1]> {
...
         } = frame;
 
         let mut log = LogEvent::default();
-        log.insert(log_schema().timestamp_key(), timestamp);
+
+        log_namespace.insert_vector_metadata(
+            &mut log,
+            log_schema().timestamp_key(),
+            "ingest_timestamp",
+            timestamp,
```
[the remark this answers is not in the document]
Comment (neuronull) [reply src/sources/fluent/mod.rs]: addressed in https://github.com/vectordotdev/vector/pull/15206/commits/eaef11960ef305327ed73ab6f31c0c6fb3e66934

File: src/sources/fluent/mod.rs (hunk 2 above)
Comment (neuronull) [reply src/sources/fluent/mod.rs]: addressed in https://github.com/vectordotdev/vector/pull/15206/commits/eaef11960ef305327ed73ab6f31c0c6fb3e66934

File: src/sources/fluent/mod.rs [hunk 4]
```diff
@@ -92,8 +101,50 @@ impl SourceConfig for FluentConfig {
...
+                host_key,
+                &owned_value_path!("host"),
+                Kind::bytes().or_undefined(),
+                Some("host"),
+            )
+            .with_source_metadata(
+                FluentConfig::NAME,
+                tag_key,
+                &owned_value_path!("tag"),
+                Kind::bytes().or_undefined(),
```
Comment (davidhuie-dd) [inline src/sources/fluent/mod.rs:130]: I missed these `_or*` functions in other PRs. Good catching finding them.

File: src/sources/fluent/mod.rs [hunk 5]
```diff
@@ -445,19 +528,37 @@ impl From<FluentFrame> for SmallVec<[Event; 1]> {
...
+            log_namespace,
         } = frame;
 
         let mut log = LogEvent::default();
-        log.insert(log_schema().timestamp_key(), timestamp);
-        log.insert("tag", tag);
+
+        match log_namespace {
+            LogNamespace::Vector => {
+                log.insert(metadata_path!(FluentConfig::NAME, "timestamp"), timestamp);
```
[the remark this answers is not in the document]
Comment (neuronull) [reply src/sources/fluent/mod.rs:544]: :thinking:  good question.
On the surface it looks like that might work. The main difference is here I'm using the source's namespace and not the "vector" namespace... I'm honestly not sure which one is correct.
I was following some direction from @spencergilbert he provided yesterday to do something similar to the changes he did in this PR
https://github.com/vectordotdev/vector/pull/15205/files
, perhaps I followed it too closely and it doesn't apply here.

File: src/sources/fluent/mod.rs [hunk 6]
```diff
@@ -92,8 +101,50 @@ impl SourceConfig for FluentConfig {
...
+                &owned_value_path!("tag"),
+                Kind::bytes().or_undefined(),
+                Some("tag"),
+            )
+            // for metadata that is added to the events dynamically from the FluentRecord
+            .with_source_metadata(
+                FluentConfig::NAME,
+                None,
+                &owned_value_path!("record"),
+                Kind::object(Collection::empty().with_unknown(Kind::bytes())),
```
Comment (spencergilbert) [inline src/sources/fluent/mod.rs:138]: If the entire `record` object may not be present (from say, no additional meta fields on the `FluentRecord`) this entire kind could be `.or_undefined()`

File: src/sources/fluent/mod.rs [hunk 7]
```diff
@@ -925,6 +1029,90 @@ mod tests {
...
+            .clone()
+            .log_schema_definition
+            .unwrap();
+
+        let expected_definition =
+            Definition::new_with_default_metadata(Kind::bytes(), [LogNamespace::Vector])
+                .with_meaning(LookupBuf::root(), "message")
+                .with_metadata_field(&owned_value_path!("vector", "source_type"), Kind::bytes())
+                .with_metadata_field(
+                    &owned_value_path!("fluent", "tag"),
```
Comment (spencergilbert) [inline src/sources/fluent/mod.rs:1055]: I'd expect this to be failing since `tag` was inserted with

```rust
        log_namespace.insert_vector_metadata(&mut log, path!("tag"), path!("tag"), tag);

```

File: src/sources/fluent/mod.rs (hunk 6 above)
Comment (neuronull) [reply src/sources/fluent/mod.rs:138]: Good call. Though I don't know much about the actual behavior of fluent, the record is a btreemap and thus could be empty and the record would be unpopulated.

File: src/sources/fluent/mod.rs (hunk 7 above)
Comment (neuronull) [reply src/sources/fluent/mod.rs:1055]: I believe the test case didn't fail because this test is just comparing the schema definition to the expected, not actual events

Comment (neuronull) [reply src/sources/fluent/mod.rs:1055]: Which I guess begs the question, do we need to have _both_ types of unit tests and not just one or the other

File: src/sources/fluent/mod.rs (hunk 4 above)
Comment (neuronull) [reply src/sources/fluent/mod.rs:130]: :+1:  glad you asked the question about these, because turns out I had not correct usage here, will update the diff.

File: src/sources/fluent/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (neuronull) [reply src/sources/fluent/mod.rs:545]: I believe I addressed this in https://github.com/vectordotdev/vector/pull/15206/commits/d39b139d770974341a022f539cde8087da4282c7

File: src/sources/fluent/mod.rs (hunk 7 above)
Comment (spencergilbert) [reply src/sources/fluent/mod.rs]: The helper @fuchsnj added can do a type comparison when you have events, so you could combine the two. I think we want to make sure, especially for `Legacy` that the shape of the events aren't changing.

File: src/sources/fluent/mod.rs [hunk 8]
```diff
@@ -92,8 +102,50 @@ impl SourceConfig for FluentConfig {
...
+                &owned_value_path!("host"),
+                Kind::bytes(),
+                Some("host"),
+            )
+            .with_source_metadata(
+                FluentConfig::NAME,
+                tag_key,
+                &owned_value_path!("tag"),
+                Kind::bytes(),
+                Some("tag"),
```
Comment (spencergilbert) [inline src/sources/fluent/mod.rs:132]:
```suggestion
                None,
```

I've only been setting a meaning, when it's `host` as we really haven't explored this anywhere except for the `datadog_agent`.

File: src/sources/fluent/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (spencergilbert) [reply src/sources/fluent/mod.rs:545]: Yep, looks good now - if the source could process batches of events at a time we could save on syscalls by caching the `now()` value across the whole batch, but we're very wasteful with `now()`s today as is 😄

File: src/sources/fluent/mod.rs [hunk 9]
```diff
@@ -925,6 +1029,90 @@ mod tests {
...
+            address: SocketListenAddr::SocketAddr("0.0.0.0:24224".parse().unwrap()),
+            tls: None,
+            keepalive: None,
+            receive_buffer_bytes: None,
+            acknowledgements: false.into(),
+            connection_limit: None,
+            log_namespace: None,
+        };
+
+        let definition = config.outputs(LogNamespace::Vector)[0]
```
[the remark this answers is not in the document]
Comment (spencergilbert) [reply src/sources/fluent/mod.rs:1053]: 🤔 without cracking open an editor, it seems strange you need to pass in this `LogNamespace` enum even though it should be accessible from your configured config.

File: src/sources/fluent/mod.rs [hunk 10]
```diff
@@ -92,8 +102,50 @@ impl SourceConfig for FluentConfig {
...
+                Some("tag"),
+            )
+            // for metadata that is added to the events dynamically from the FluentRecord
+            .with_source_metadata(
+                FluentConfig::NAME,
+                None,
+                &owned_value_path!("record"),
+                Kind::object(Collection::empty().with_unknown(Kind::bytes())).or_undefined(),
+                None,
+            );
```
Comment (spencergilbert) [inline src/sources/fluent/mod.rs:141]: Oops, just noticed this is missing the new `fluent.timestamp` metadata value that gets added in the Vector namespace!

**Review by spencergilbert — CHANGES_REQUESTED** [review]

File: src/sources/fluent/mod.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (neuronull) [reply src/sources/fluent/mod.rs:1053]: The `outputs()` function takes a global_log_namespace argument.

File: src/sources/fluent/mod.rs (hunk 8 above)
Comment (neuronull) [reply src/sources/fluent/mod.rs:132]: ok, I'll adapt to that convention. Not exactly clear on what "meaning" ... means :laughing:

File: src/sources/fluent/mod.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (spencergilbert) [reply src/sources/fluent/mod.rs:1053]: Word, destined to just look wack forever 😆

File: src/sources/fluent/mod.rs (hunk 10 above)
Comment (neuronull) [reply src/sources/fluent/mod.rs:141]: hummm.. are `amqp` and `aws_sqs` also missing that me thinks?

File: src/sources/fluent/mod.rs (hunk 8 above)
Comment (spencergilbert) [reply src/sources/fluent/mod.rs:132]: Yeah, I'd say it's not clearly defined at any point - so not something you're really "missing"

File: src/sources/fluent/mod.rs (hunk 10 above)
Comment (neuronull) [reply src/sources/fluent/mod.rs:164]: Added to all 3 places in https://github.com/vectordotdev/vector/pull/15206/commits/d1b11ad058d9e44d70df9b740a2bbbc045b97c14

**Review by spencergilbert — APPROVED** [review]

**Review by davidhuie-dd — APPROVED** [review]

File: website/content/en/highlights/2022-11-07-0-26-0-upgrade-guide.md [hunk 11]
```diff
@@ -14,6 +14,7 @@ Vector's 0.26.0 release includes **breaking changes**:
 1. [Removal of `check_fields` condition type](#check-fields-removal)
 2. [Removal of `lanes` parameter for `route` transform](#lanes-parameter-route-transform-removal)
 3. [Removal of `geoip` transform](#geo-transform-removal)
+3. [Enrichment of `logstash` source with `source_type` field](#logstash-source-type-field)
```
Comment (spencergilbert) [inline website/content/en/highlights/2022-11-07-0-26-0-upgrade-guide.md:17]:
```suggestion
3. [Enrichment of `fluent` source with `source_type` field](#logstash-source-type-field)
```
?

File: website/content/en/highlights/2022-11-07-0-26-0-upgrade-guide.md [hunk 12]
```diff
@@ -66,3 +67,11 @@ get the same functionality through a `remap` transform with [`geoip` enrichment
...
 with [VRL's enrichment table functions](/docs/reference/vrl/functions/#enrichment-functions) to
 enrich events using a [GeoIP database](https://www.maxmind.com/en/geoip2-databases).
+
+#### Enrichment of `logstash` source with `source_type` field {#logstash-source-type-field}
+
+As part of the effort to add log namespacing support to all sources, in addition to setting the
+`source_type` to the Vector namespace, the Legacy namespace was also modified to enrich events
+with the global option [log_schema.source_type_key](https://vector.dev/docs/reference/configuration/global-options/#log_schema.source_type_key).
+
+This marks a breaking change to the shape of the log events output from the `logstash` source.
```
Comment (spencergilbert) [inline website/content/en/highlights/2022-11-07-0-26-0-upgrade-guide.md:77]: I'd rewrite this to not actually mention the legacy vs vector namespace - we've not announced/documented/etc these yet so just saying something about including this field for consistency and all that jazz

File: website/content/en/highlights/2022-11-07-0-26-0-upgrade-guide.md (hunk 11 above)
Comment (neuronull) [reply website/content/en/highlights/2022-11-07-0-26-0-upgrade-guide.md:17]: :eyes:

File: src/sources/amqp.rs [hunk 13]
```diff
@@ -135,7 +136,14 @@ impl SourceConfig for AmqpSourceConfig {
         let schema_definition = self
             .decoding
             .schema_definition(log_namespace)
-            .with_standard_vector_source_metadata();
+            .with_standard_vector_source_metadata()
+            .with_source_metadata(
```
Comment (StephenWakely) [inline src/sources/amqp.rs:140]: This isn't the fluent source... Should we put this in a separate pr?

Comment (spencergilbert) [reply src/sources/amqp.rs:140]: He had noticed I had incorrectly not included this type def in the amqp and aws_s3 fixes I PR'd earlier. I don't have a problem including it in this PR, but if there's a group preference for separating 👍

Comment (neuronull) [reply src/sources/amqp.rs:140]: Ditto to Spencer's response.
We had chatted offline about this as I observed this missing from the schema def in this PR (after it was merged in) https://github.com/vectordotdev/vector/pull/15205

The relevant lines where that field is added are here :
https://github.com/vectordotdev/vector/blob/6099e0d3eace517d1a6b2cda5daab61c70a89ac2/src/sources/amqp.rs#L234-L255

If you feel strongly about it being pulled out, I can do that.

Main reason I didn't was just that it's a pretty small change. In hindsight I should have added a note about it to the PR description.

Comment (StephenWakely) [reply src/sources/amqp.rs:140]: It's not a huge issue. It's just I had also added this to the amqp source, but haven't submitted a PR for it yet so one of us will get a conflict.. but not really a big deal.

File: website/content/en/highlights/2022-11-07-0-26-0-upgrade-guide.md [hunk 14]
```diff
@@ -66,3 +67,11 @@ get the same functionality through a `remap` transform with [`geoip` enrichment
 tables](/docs/reference/configuration/global-options/#enrichment_tables.geoip). These can be used
 with [VRL's enrichment table functions](/docs/reference/vrl/functions/#enrichment-functions) to
 enrich events using a [GeoIP database](https://www.maxmind.com/en/geoip2-databases).
+
+#### Enrichment of `fluent` source with `source_type` field {#fluent-source-type-field}
```
Comment (jszwedko) [inline website/content/en/highlights/2022-11-07-0-26-0-upgrade-guide.md:71]: We typically haven't considered field additions to be breaking changes. I think we can drop this note.

Comment (neuronull) [reply website/content/en/highlights/2022-11-07-0-26-0-upgrade-guide.md:71]: Oh good to know... I also made changes to this file for the `vector` source PR. So I can remove that in that PR.
